### PR TITLE
AIRAVATA-3310: Enabling file edit at the experiments

### PR DESCRIPTION
**Pull request by dinukadesilva**

### CODE REVIEW

**Comment by dinukadesilva** [comment]
@machristie @eroma2014  This PR is ready for review

File: django_airavata/apps/api/views.py [hunk 1]
```diff
@@ -970,6 +970,16 @@ def get(self, request, format=None):
...
         return Response(serializer.data)
 
+    def put(self, request, format=None):
+        data_product_uri = request.query_params['product-uri']
+        data_product = request.airavata_client.getDataProduct(
+            request.authz_token, data_product_uri)
+
+        return UserStoragePathView().put(
+            request=request,
+            path=data_product.replicaLocations[0].filePath.replace("file://localhost:/tmp/default-admin/", "")
```
Comment (machristie) [inline django_airavata/apps/api/views.py:980]: `file://localhost:/tmp/default-admin/` only works based on your local setup, it doesn't generally work. You need a more reliable way to compute the user relative path to the file.

`user_storage.get_rel_path` can return a relative path for a full path. https://github.com/apache/airavata-django-portal-sdk/blob/master/airavata_django_portal_sdk/user_storage.py#L499

However, we need to also implement this for the remote API mode of user_storage, but that can be via a separate issue/PR.

File: django_airavata/apps/workspace/static/django_airavata_workspace/js/containers/UserStorageContainer.vue [hunk 2]
```diff
@@ -10,43 +10,60 @@
...
+      }
+
+      this.dataProductUri = _dataProductUri
+    },
+    async setStoragePath() {
+      this.setDataProductUri();
+      let _storagePath = null;
+      if (this.dataProductUri) {
+        const dataProduct = await utils.FetchUtils.get(`/api/data-products?product-uri=${this.dataProductUri}`);
+        _storagePath = dataProduct.replicaLocations[0].filePath.replace("file://localhost:/tmp/default-admin/", "~/")
```
Comment (machristie) [inline django_airavata/apps/workspace/static/django_airavata_workspace/js/containers/UserStorageContainer.vue:46]: likewise here too

File: django_airavata/apps/workspace/static/django_airavata_workspace/js/containers/UserStorageContainer.vue [hunk 3]
```diff
@@ -10,43 +10,60 @@
...
+  data() {
+    return {
+      dataProductUri: null,
+      storagePath: null,
+      userStoragePath: null
+    };
+  },
+  methods: {
+    setDataProductUri() {
+      let _dataProductUri = /\?.*dataProductUri=(.*)/.exec(window.location.href);
```
Comment (machristie) [inline django_airavata/apps/workspace/static/django_airavata_workspace/js/containers/UserStorageContainer.vue:34]: Why is this needed? This seems like a weird way to get the data product URI.

Comment (machristie) [reply django_airavata/apps/workspace/static/django_airavata_workspace/js/containers/UserStorageContainer.vue:34]: Okay, I think I see. Keep in mind you can pass props to routes, that might be a better way to pass the dataProductUri https://router.vuejs.org/guide/essentials/passing-props.html. In fact you can keep the old behavior of the routing being based on the path but pass the dataProductUri as a prop.

**Review by machristie — CHANGES_REQUESTED** [review]
Hi @dinukadesilva , looks good, I just have a few requested changes. Let me know if you have questions. Thanks.

File: django_airavata/apps/api/views.py (hunk 1 above)
Comment (dinukadesilva) [reply django_airavata/apps/api/views.py]: Actually user_storage.get_rel_path didn't do that. Instead, I moved this to the serializer and improved the regex

File: django_airavata/apps/workspace/static/django_airavata_workspace/js/containers/UserStorageContainer.vue (hunk 3 above)
Comment (dinukadesilva) [reply django_airavata/apps/workspace/static/django_airavata_workspace/js/containers/UserStorageContainer.vue]: Changed it to `this.$route.query.dataProductUri`

File: django_airavata/apps/api/serializers.py [hunk 4]
```diff
@@ -513,6 +515,13 @@ def get_isInputFileUpload(self, data_product):
         request = self.context['request']
         return user_storage.is_input_file(request, data_product)
 
+    def get_path(self, data_product):
+        """Getter for path field."""
+        if len(data_product.replicaLocations) > 0:
+            return re.sub(r'.*/tmp/[^/]*/', "", data_product.replicaLocations[0].filePath)
```
Comment (machristie) [inline django_airavata/apps/api/serializers.py:521]: This won't work because not all locations are in /tmp/. That's just kind of a local development setup. I think we just need to treat data products differently than paths in user storage.

File: django_airavata/apps/api/views.py [hunk 5]
```diff
@@ -974,10 +974,12 @@ def put(self, request, format=None):
         data_product_uri = request.query_params['product-uri']
         data_product = request.airavata_client.getDataProduct(
             request.authz_token, data_product_uri)
+        serializer = self.serializer_class(
+            data_product, context={'request': request})
 
         return UserStoragePathView().put(
```
Comment (machristie) [inline django_airavata/apps/api/views.py:980]: Instead of calling `UserStoragePathView().put()`, what if we had either a different version of `user_storage.update_file_content` (maybe called `update_data_product_content`?) or one that took an optional `data_product` argument that would figure out the path internally. `user_storage` already has a `_get_replica_filepath` function - https://github.com/apache/airavata-django-portal-sdk/blob/master/airavata_django_portal_sdk/user_storage.py#L624.

File: django_airavata/apps/workspace/static/django_airavata_workspace/js/components/storage/storage-edit/UserStorageLink.vue [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  methods: {
+    showFilePreview(event) {
+      if (this.allowPreview) {
+        this.$refs.modal.show();
+        event.preventDefault();
+      }
+    },
+    storageFileViewRouteUrl() {
+      // This endpoint can handle XHR upload or a TUS uploadURL
+      return `/workspace/storage/~?dataProductUri=${this.dataProductUri}`;
```
Comment (machristie) [inline django_airavata/apps/workspace/static/django_airavata_workspace/js/components/storage/storage-edit/UserStorageLink.vue:50]: It would be nice if we could keep the regular URL path and pass the dataProductUri as a prop instead.

**Review by machristie — CHANGES_REQUESTED** [review]
Thanks @dinukadesilva . I've added some suggestions.

Also, please look at the Travis build output. I think some of the static analysis is failing for the PR changes.

File: django_airavata/apps/api/serializers.py (hunk 4 above)
Comment (dinukadesilva) [reply django_airavata/apps/api/serializers.py:521]: @machristie Do you have any thoughts on how to do that. Currently, I cannot see any existing utilities or any other way.

File: django_airavata/apps/workspace/static/django_airavata_workspace/js/components/storage/storage-edit/UserStorageLink.vue (hunk 6 above)
Comment (dinukadesilva) [reply django_airavata/apps/workspace/static/django_airavata_workspace/js/components/storage/storage-edit/UserStorageLink.vue:50]: @machristie dataProductUri  is a prop already passed to this component

https://github.com/apache/airavata-django-portal/pull/52/files/3c049283cb5211f76c4576251ffb59e55afd7a52#diff-589038988d47ba2be64cd940ebaa158b28630aa899bf0104ab66d9c69b09b2a9R30

File: django_airavata/apps/api/serializers.py (hunk 4 above)
Comment (machristie) [reply django_airavata/apps/api/serializers.py:521]: Yeah, see my other comment below. Essentially, if we have a data product URI, that should be all that the API needs. The API shouldn't need the path because the path can be determined from the data product URI. I'll create an updated `user_storage.update_file_content` function in the SDK that you can use in DataProductView.put()

File: django_airavata/apps/workspace/static/django_airavata_workspace/js/components/storage/storage-edit/UserStorageLink.vue (hunk 6 above)
Comment (machristie) [reply django_airavata/apps/workspace/static/django_airavata_workspace/js/components/storage/storage-edit/UserStorageLink.vue:50]: No I'm talking about the link produced here. The link should be the path to the file, like before https://github.com/apache/airavata-django-portal/blob/develop/django_airavata/apps/workspace/static/django_airavata_workspace/js/components/storage/UserStoragePathViewer.vue#L209.

Well, I guess, I don't quite understand why the dataProductUri needs to be a query parameter. Looks like UserStorageContainer uses it to get the path, but it wouldn't need that if the path were used like before. Can you explain why dataProductUri is needed as a query parameter?

File: django_airavata/apps/api/serializers.py (hunk 4 above)
Comment (machristie) [reply django_airavata/apps/api/serializers.py:521]: @dinukadesilva I've added an update_data_product_content function to the user_storage.py module - https://github.com/apache/airavata-django-portal-sdk/blob/master/airavata_django_portal_sdk/user_storage.py#L268.

To use the latest sdk, you'll need to clone it and then install it into your airavata-django-portal virtual environment
```
# activate your airavata-django-portal venv
cd airavata-django-portal
source venv/bin/activate
cd ..

# clone and install SDK into portal venv
git clone https://github.com/apache/airavata-django-portal-sdk.git
cd airavata-django-portal-sdk
python setup.py develop
```

Comment (dinukadesilva) [reply django_airavata/apps/api/serializers.py]: This worked @machristie . Thanks

File: django_airavata/apps/workspace/static/django_airavata_workspace/js/components/storage/storage-edit/UserStorageLink.vue (hunk 6 above)
Comment (dinukadesilva) [reply django_airavata/apps/workspace/static/django_airavata_workspace/js/components/storage/storage-edit/UserStorageLink.vue:50]: @machristie Passing dataProductUri as a query parameter is only an additional thing and passing the path is still working as well. I did so because I do not have the path in the URL when the file editor is opened through a dialog.

File: django_airavata/apps/api/serializers.py [hunk 7]
```diff
@@ -513,6 +515,13 @@ def get_isInputFileUpload(self, data_product):
         request = self.context['request']
         return user_storage.is_input_file(request, data_product)
 
+    def get_path(self, data_product):
+        """Getter for path field."""
+        if len(data_product.replicaLocations) > 0:
+            return re.sub(r'.*/tmp/[^/]*/', "", user_storage._get_replica_filepath(data_product))
```
Comment (dinukadesilva) [inline django_airavata/apps/api/serializers.py:521]: @machristie Can you help me with a workaround to make this generic

File: django_airavata/apps/api/views.py (hunk 5 above)
Comment (dinukadesilva) [reply django_airavata/apps/api/views.py]: This worked @machristie 
Thanks

File: django_airavata/apps/api/serializers.py (hunk 7 above)
Comment (machristie) [reply django_airavata/apps/api/serializers.py:521]: Is this still needed? With your most recent commit, only the data product uri is needed to update the file contents.

**Comment by machristie** [comment]
@dinukadesilva the travis build is failing, probably some static analysis failure. You might want to merge/rebase with develop since some of the errors may have been fixed on develop. Then can you look into the build errors and fix them?

File: django_airavata/apps/api/serializers.py (hunk 7 above)
Comment (dinukadesilva) [reply django_airavata/apps/api/serializers.py:521]: It is because from the frontend, first I take the data product from the data product uri and using the path in it only the requests to UserStoragePathView are done. UserStoragePathView requires the path.

Comment (machristie) [reply django_airavata/apps/api/serializers.py:521]: Can you point out where that is in the code? Thanks.

Comment (dinukadesilva) [reply django_airavata/apps/api/serializers.py:521]: https://github.com/apache/airavata-django-portal/pull/52/files#diff-c7351d24aefb0b542f4eb165ac38b5271701a7fd111e44ed45614e181a4df817R59

Comment (dinukadesilva) [reply django_airavata/apps/api/serializers.py:521]: Sure. Here's in frontend [1] and backend [2] places where the path is required.

[1] https://github.com/dinukadesilva/airavata-django-portal/blob/AIRAVATA-3310-enabling-file-edit-at-the-experiments/django_airavata/apps/workspace/static/django_airavata_workspace/js/containers/UserStorageContainer.vue#L59

[2] https://github.com/dinukadesilva/airavata-django-portal/blob/AIRAVATA-3310-enabling-file-edit-at-the-experiments/django_airavata/apps/api/views.py#L1542

Comment (machristie) [reply django_airavata/apps/api/serializers.py:512]: @dinukadesilva those links don't seem to be working any longer, meaning they don't point to the right lines. But I tried removing the `path` field from the DataProductSerializer and I was able to upload an input file and then edit it. So I don't think it is needed any longer. Try removing it and see if it works for you too.

Comment (dinukadesilva) [reply django_airavata/apps/api/serializers.py:512]: @machristie Actually the path is needed in the following two
1. If you go to storage and click on one of the files
2. You go to an application, select a file, right-click on the link and open in a new tab.

It's because the "/api/user-storage/~/$path" endpoint needs the path to be given as a param. Earlier this was taken from the window URL. But now we had to change it since the window URL is not yet valid when it's at the experiments window.

Comment (machristie) [reply django_airavata/apps/api/serializers.py:512]: Neither of those work for me when I try the code locally.  Conceptually, from the API you should only need to provide the data product uri to edit a file, the path shouldn't be needed. The backend can take care of figuring out where the file is actually located.

Maybe we should meet to figure out a way forward.

**Comment by machristie** [comment]
Thanks @dinukadesilva, it looks great!
